crates/derive/src/combine_api.rs

```diff
@@ -225,6 +236,7 @@ pub fn drain_chunk(
     arena: &mut Vec<u8>,
     out: &mut Vec<cgo::Out>,
     stats: &mut DocCounter,
+    shape: Option<&doc::inference::Shape>,
```

Comment:
I'm thinking `shape` should be required here? 🤔 (though this can wait for a future PR). 

In practice, when `drain_chunk` is used in derivations, we use it to extract collection keys and logically partitioned values. Currently we _require_ that those locations exist, so a default value will never be used, and it doesn't _currently_ matter.

But having started down this road, I'm not sure why we _wouldn't_ allow collection key components or logically partitioned values to be omitted so long as they have valid default values.

As a practical benefit, that would allow a user to extend the key of a collection with a new component which might only be present in some documents (say, because it was recently added).

Reply:
I think I can see how we could set up the derivation pipeline to provide a `shape` pretty easily. Making use of it the way you describe would take more effort though, so I think I'll leave it as an option & the derive pipeline providing `None` for now until we want to tackle default values for derivations.